File: backend/java-spring/api/TasksController.java
package cheat.tasks;

// Why this file exists:
// - Minimal endpoint shapes (working, but simple)

import org.springframework.web.bind.annotation.*;
import java.util.List;

@RestController
public class TasksController {
  private final TaskRepo repo;
  private final RabbitPublisher publisher;

  public TasksController(TaskRepo repo, RabbitPublisher publisher) {
    this.repo = repo;
    this.publisher = publisher;
  }

  @GetMapping("/health")
  public Object health() { return new Object() { public boolean ok = true; }; }

  @GetMapping("/tasks")
  public List<TaskItem> tasks() {
    return repo.findAll();
  }

  @PostMapping("/tasks")
  public TaskItem create(@RequestBody TaskItem body) {
    body.done = false;
    TaskItem saved = repo.save(body);

    publisher.publishTaskCreated(saved.title);
    return saved;
  }

  @PatchMapping("/tasks/{id}/done")
  public Object toggleDone(@PathVariable Long id) {
    var taskOpt = repo.findById(id);
    if (taskOpt.isEmpty()) return new Object() { public String error = "not found"; };

    var task = taskOpt.get();
    task.done = !task.done;
    repo.save(task);

    return task;
  }
}

File: backend/nestjs/api/tasks.controller.ts
// Why this file exists:
// - Minimal Nest controller + minimal ORM usage (working, but simple)

import { Controller, Get, Post, Patch, Param, Body } from "@nestjs/common";
import { InjectRepository } from "@nestjs/typeorm";
import { Repository } from "typeorm";
import { TaskItem } from "./task.entity";
import { publishTaskCreated } from "./rabbit.producer";

@Controller()
export class TasksController {
  constructor(
    @InjectRepository(TaskItem) private tasks: Repository<TaskItem>,
  ) {}

  @Get("health")
  health() {
    return { ok: true };
  }

  @Get("tasks")
  async list() {
    return this.tasks.find();
  }

  @Post("tasks")
  async create(@Body() body: any) {
    const title = body?.title ?? "untitled";
    const task = this.tasks.create({ title, done: false });
    const saved = await this.tasks.save(task);

    await publishTaskCreated(title);
    return saved;
  }

  @Patch("tasks/:id/done")
  async toggle(@Param("id") id: string) {
    const task = await this.tasks.findOneBy({ id: Number(id) });
    if (!task) return { error: "not found" };

    task.done = !task.done;
    return this.tasks.save(task);
  }
}
